clients/src/main/java/org/apache/kafka/common/requests/ListGroupsRequest.java

```
@@ -50,8 +53,19 @@ public ListGroupsRequest build(short version) {
                         "v" + version + ", but we need v4 or newer to request groups by states.");
             }
             if (!data.typesFilter().isEmpty() && version < 5) {
-                throw new UnsupportedVersionException("The broker only supports ListGroups " +
-                    "v" + version + ", but we need v5 or newer to request groups by type.");
+                // Types filter is supported by brokers with version 4.0.0 or later. Older brokers only support
+                // classic groups, so listing consumer groups on an older broker does not need to use a types filter.
+                // If the types filter is only for consumer and classic, or just classic groups, it can be safely omitted.
+                // This allows a modern admin client to list consumer groups on older brokers in a straightforward way.
+                HashSet<String> typesCopy = new HashSet<>(data.typesFilter());
+                boolean containedClassic = typesCopy.remove(GroupType.CLASSIC.toString());
+                boolean containedConsumer = typesCopy.remove(GroupType.CONSUMER.toString());
+                if (!typesCopy.isEmpty() || (!containedClassic && containedConsumer)) {
```

The KIP-848 was shipped with 3.7 as early access. Is it possible that 3.7 broker return both classic and consumer groups even though the LIST_GROUP requests only classic group?

---

I do know that `ListGroupsRequest.TypesFilter` and `ListGroupsResponse.GroupType` were introduced together and, as you correctly pointed out, they were introduced in 3.8.

My point is that if you connect to a broker which is too old to support the TypesFilter and GroupType, you cannot control which types of groups it returns, nor distinguish between different types of group. It's classic groups only.

For a pre-3.8 broker:
* `ListGroupsOptions.withTypes(List.of(GroupType.CLASSIC, GroupType.CONSUMER))` will return classic groups. This is the same as `ListGroupsOptions.forConsumerGroups())`. The lack of TypesFilter is no problem.
* `ListGroupsOptions.withTypes(List.of(GroupType.CLASSIC))` will return classic groups, because it doesn't know any other types. The lack of TypesFilter is no problem.
* `ListGroupsOptions.withType(List.of(GroupType.CONSUMER))` will throw `UnsupportedVersionException`. The user has explicitly requested only `GroupType.CONSUMER` and that cannot be expressed without a TypesFilter.

There is a tiny outlier for 3.7 brokers with KIP-848 EA enabled. Even those didn't support TypesFilter or GroupType. All groups would have appeared to be classic groups. So, I think the above 3 cases are as good as can be expected.  Modern consumer groups would be reported as classic groups, because the RPC response was not able to distinguish.

---

> Modern consumer groups would be reported as classic groups, because the RPC response was not able to distinguish.

My point was the request `ListGroupsOptions.withTypes(List.of(GroupType.CLASSIC))` will get both classic and consumer groups from 3.7 broker when KIP-848 EA enabled. That is a bit weird to me, as users are expected to receive only the classic groups. Hence, maybe we should throw `UnsupportedVersionException` for that case?

---

I'm afraid I don't agree. For a 3.7 broker with KIP-848 EA enabled, all groups appear to be classic groups. Because the KIP is incomplete, the broker can neither filter by group type, nor respond with the correct group type. All groups appear to be classic groups. So, I feel this is an outlier specifically because of the partial implementation of the KIP. In 3.8, more of KIP-848 was implemented and it supported the newer RPC version.

If the user asks for `ListGroupsOptions.withTypes(List.of(GroupType.CLASSIC))` in all cases except 3.7 with KIP-848 EA enabled, it's going to do the right thing. I would rather not throw `UnsupportedVersionException` for these working cases because of the KIP-848 EA case.

---

Got it. Thanks for the explanation. +1